### PR TITLE
Add Karp's minimum cycle mean algorithm

### LINKED ISSUE #512: Maximum/Minimum cycle mean of weighted digraph
Is [Karp's algorithm](http://www.sciencedirect.com/science/article/pii/0012365X78900110) or [faster](http://ieeexplore.ieee.org/document/728912/?denied) implemented for Max/Minimum cycle mean on weighted digraph ?
If not, in which package of `JuliaGraphs` should this be added ?

### PR DESCRIPTION
**Pull request by blegat**
Fixes #512

### CODE REVIEW

**Comment by blegat** [comment]
I get the following failure
```julia
Diffusion Simulation: Test Failed
  Expression: avg > means[p] - (stds[p] / sqrt(runs)) * 3.5
   Evaluated: 0.8999999999999999 > 1.0100505063388334
```
However, I don't see how the PR have affected this. Is it a known failure ?

**Comment by sbromberger** [comment]
>  Is it a known failure ?

Yes - it happens occasionally. I'll merge and retest.

File: src/digraph/cycles/karp.jl [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+### References
+- [Karp](http://dx.doi.org/10.1016/0012-365X(78)90011-0).
+"""
+function karp_minimum_cycle_mean end
+#@traitfn function karp_minimum_cycle_mean(
+#    g::::IsDirected,
+#    distmx::::AbstractMatrix = weights(g)
+#    )
+function karp_minimum_cycle_mean(g, distmx)
```
Comment (sbromberger) [inline src/digraph/cycles/karp.jl]: Make sure `distmx` has a default:
```
distmx::AbstractMatrix = weights(g);
```
(also change the docstring:
```
"""
    karp_minimum_cycle_mean(g[, distmx])

Return minimum cycle mean of the directed graph `g` with optional edge weights contained in `distmx`.
"""

File: src/digraph/cycles/karp.jl [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        λ = maximum(map(i -> (F[n+1, j] - F[i, j]) / (n+1 - i), 1:n))
+        if λ < λmin || (isfinite(λ) && λ ≈ λmin && F[n+1, j] < F[n+1, jbest])
+            λmin = λ
+            jbest = j
+        end
+    end
+    
+    if iszero(jbest)
+        return U[], Inf
+    end
```
Comment (sbromberger) [inline src/digraph/cycles/karp.jl]: Short-circuit here:
```
iszero(jbest) && return U[], Inf
```

File: src/digraph/cycles/karp.jl [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    v2j = Dict{U, Int}()
+    for (j, v) in enumerate(component)
+        v2j[v] = j
+    end
+    n = length(component)
+    F = Matrix{float(T)}(n+1, n)
+    F[1, 1] = 0.
+    for v in 2:length(component)
+        F[1, v] = Inf
+    end
```
Comment (sbromberger) [inline src/digraph/cycles/karp.jl]: can you use `fill` here somehow?

File: src/digraph/cycles/karp.jl [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    distmx::AbstractMatrix{T},
+    component::Vector{U}
+    ) where T where U<:Integer
+
+    v2j = Dict{U, Int}()
+    for (j, v) in enumerate(component)
+        v2j[v] = j
+    end
+    n = length(component)
+    F = Matrix{float(T)}(n+1, n)
```
Comment (sbromberger) [inline src/digraph/cycles/karp.jl]: Does it make sense just to make this `Float64`?

File: src/digraph/cycles/karp.jl [hunk 5]
```diff
@@ -0,0 +1,119 @@
+# Karp, R. M.
+# A characterization of the minimum cycle mean in a digraph
+# Discrete Mathematics, 1978, 23, 309 - 311
+function _karp_minimum_cycle_mean(
+    g::AbstractGraph,
+    distmx::AbstractMatrix{T},
+    component::Vector{U}
+    ) where T where U<:Integer
```
Comment (sbromberger) [inline src/digraph/cycles/karp.jl]: `where T <: Real`

File: test/digraph/cycles/karp.jl [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+     add_edge!(g, 6, 2)
+     add_edge!(g, 6, 7)
+     add_edge!(g, 7, 2)
+     add_edge!(g, 7, 9)
+     add_edge!(g, 9, 11)
+     add_edge!(g, 10, 7)
+     add_edge!(g, 11, 7)
+     add_edge!(g, 12, 9)
+
+     c, λ = karp_minimum_cycle_mean(g, w)
```
Comment (sbromberger) [inline test/digraph/cycles/karp.jl]: make sure you cycle through all graph types: rename `g` to `g1` and do
```
for g in maketestgraphs(g1)
  c, λ = @inferred(karp_minimum_cycle_mean(g, w))
     @test c == [9, 11, 7]
     @test λ == 0.9
end
```

(repeat for the other tests as well)

**Comment by blegat** [comment]
Thanks for the comments ! I have improved the PR

**Comment by sbromberger** [comment]
Great. I'll do another quick review and would like @jpfairbanks to check it for functionality and things I've missed. :)

**Comment by sbromberger** [comment]
awesome. Thank you, @blegat and @jpfairbanks !!

**Pull request merged**